### DISCUSSION TITLE
BR fails to send RA

### DISCUSSION THREAD

**Discussion by pptz** [openthread/openthread#10739] [Q&A] [answered]
I am facing a rather odd issue with an OT border router implementation of my own making on an ESP32 platform. According to the logs, the BR advertises its NAT64 prefix on the WiFi LAN as it should, but I don't observe the RA in a sniffer and am not able to reach end devices connected to the BR over ipv6. A reference implementation of an ESP32 BR works as expected while producing virtually the same logs. Here is a comparison of the logs between my device and the example BR (https://github.com/espressif/esp-idf/tree/master/examples/openthread/ot_br).

1. Reference:
```
> ipaddr

fd00:db8:a0:0:0:ff:fe00:fc10
fd0c:d918:2dce:1:d73b:a341:507d:4b1d
fd00:db8:a0:0:0:ff:fe00:fc00
fd00:db8:a0:0:0:ff:fe00:d000
fd00:db8:a0:0:84d:8ac9:1aaf:eeb1
fe80:0:0:0:b050:9:4184:f76b
Done

> br nat64prefix

Local: fd0c:d918:2dce:2:0:0::/96
Favored: fd0c:d918:2dce:2:0:0::/96 prf:low
Done

I(204194) OPENTHREAD:[I] RoutingManager: Preparing RA
I(204204) OPENTHREAD:[I] RoutingManager: - RA Header - flags - M:0 O:0
I(204204) OPENTHREAD:[I] RoutingManager: - RA Header - default route - lifetime:0
I(204214) OPENTHREAD:[I] RoutingManager: - FlagsExt - StubRouter:1
I(204234) OPENTHREAD:[I] RoutingManager: - RIO fd0c:d918:2dce:1::/64 (lifetime:1800, prf:medium)
I(204234) OPENTHREAD:[I] RoutingManager: Sent RA on infra netif 3
```

2. My BR:
```
> ipaddr

fd2f:6157:afae:1:6d1e:d858:5e2c:5bc5
fd12:d3e3:791b:bb06:0:ff:fe00:fc10
fd12:d3e3:791b:bb06:0:ff:fe00:fc00
fd12:d3e3:791b:bb06:0:ff:fe00:7c00
fd12:d3e3:791b:bb06:76f1:845b:f05c:b8e7
fe80:0:0:0:c63:703f:a83a:f0a3
Done

> br nat64prefix

Local: fd2f:6157:afae:2:0:0::/96
Favored: fd2f:6157:afae:2:0:0::/96 prf:low
Done

I(686713) OPENTHREAD:[I] RoutingManager: Preparing RA
I(686713) OPENTHREAD:[I] RoutingManager: - RA Header - flags - M:0 O:0
I(686723) OPENTHREAD:[I] RoutingManager: - RA Header - default route - lifetime:0
I(686723) OPENTHREAD:[I] RoutingManager: - FlagsExt - StubRouter:1
I(686743) OPENTHREAD:[I] RoutingManager: - PIO fd9d:e807:3acd:1dd7::/64 (valid:1800, preferred:1800)
I(686743) OPENTHREAD:[I] RoutingManager: - RIO fd2f:6157:afae:1::/64 (lifetime:1800, prf:medium)
I(686753) OPENTHREAD:[I] RoutingManager: Sent RA on infra netif 3
```
-----------

The only difference seems to be that in my implementation the RA also has a PIO option. I tried removing (that is, not adding) it programmatically for the sake of the experiment, it makes no difference.

Checking the `br counters` reveals that `RA TxSuccess` is a growing positive integer in both cases.

I'm currently digging into the underlying networking stack of the ESP device to see if there's some issue with it, but in the meanwhile thought I'd post here in case anyone might offer some additional ideas.

Thanks.

**Comment by abtink** [comment]
OT `RoutingManager` upon constructing the RA message will pass it to platform layer to be sent. In particular, the following `otPlatInfraIf` API will be invoked:

https://github.com/openthread/openthread/blob/9f6921cf04b277ba7a6e82260f7a6ceaebf6c36e/include/openthread/platform/infra_if.h#L70-L92

May be helpful to  check the implementation of this API on the platform.

**Comment by chshu** [comment]
@pptz The `otPlatInfraIfSendIcmp6Nd()` function is implemented and provided in the library, so the source code is not available in ESP-IDF. Do you have any specific requirement to implement the BR feature yourself, rather than using the ESP Thread BR SDK directly?

**Reply by pptz** [reply]
> @pptz The `otPlatInfraIfSendIcmp6Nd()` function is implemented and provided in the library, so the source code is not available in ESP-IDF. Do you have any specific requirement to implement the BR feature yourself, rather than using the ESP Thread BR SDK directly?

I _am_ using the ESP Thread BR API, as far as I can tell I'm calling the same methods as the reference `ot_br` ESP example, which, as I mentioned, works well. But when integrated into my product, something goes awry. I suspect that it's indeed something platform-specific, not in OT itself.

**Reply by chshu** [reply]
Could you share the steps or sample code to reproduce the issue?

**Comment by pptz** [comment, marked as answer]
I'll post the resolution here for the record, even though, as suspected, the problem had nothing to do with OpenThread itself.

Turned out that the LAN interface on my device did not have an IPv6 address. This logically broke route advertising on the LAN (as IPv6 routes are advertised with ICMPv6 using IPv6 addresses), but without any visible errors on the side of the Espressif port of OT.

**Reply by abtink** [reply]
Thanks for posting the resolution. 👍
